examples/cluster/plot_optics.py

```diff
@@ -0,0 +1,127 @@
+import matplotlib.gridspec as gridspec
+
+
+import numpy as np
+
+import matplotlib.pyplot as plt
+
+# Generate sample data
+
+np.random.seed(0)
+n_points_per_cluster = 250
+
+C1 = [-5, -2] + .8 * np.random.randn(n_points_per_cluster, 2)
+C2 = [4, -1] + .1 * np.random.randn(n_points_per_cluster, 2)
+C3 = [1, -2] + .2 * np.random.randn(n_points_per_cluster, 2)
+C4 = [-2, 3] + .3 * np.random.randn(n_points_per_cluster, 2)
+C5 = [3, -2] + 1.6 * np.random.randn(n_points_per_cluster, 2)
+C6 = [5, 6] + 2 * np.random.randn(n_points_per_cluster, 2)
+X = np.vstack((C1, C2, C3, C4, C5, C6))
+
+clust = OPTICS(min_samples=9, rejection_ratio=0.5)
+
+# Run the fit
+clust.fit(X)
+
+_, labels_025 = clust.extract_dbscan(0.25)
+_, labels_075 = clust.extract_dbscan(0.75)
+
+space = np.r_[0:len(X):1]
+R = clust.reachability_[clust.ordering_]
+C = clust.labels_[clust.ordering_]
+
+plt.figure(figsize=(10, 7))
+G = gridspec.GridSpec(2, 3)
+ax1 = plt.subplot(G[0, :])
+ax1.set_ylabel('Reachability (epsilon distance)')
+ax1.set_title('Reachability Plot')
+ax2 = plt.subplot(G[1, 0])
+ax2.set_title('Automatic Clustering')
+ax3 = plt.subplot(G[1, 1])
+ax3.set_title('Clustering at 0.25 epsilon cut')
+ax4 = plt.subplot(G[1, 2])
+ax4.set_title('Clustering at 0.75 epsilon cut')
+
+# Reachability plot
+color = ['g.', 'r.', 'b.', 'y.', 'c.']
+for k, c in zip(range(1, 7), color):
+    Xk = space[C == k]
+    Rk = R[C == k]
+    ax1.plot(Xk, Rk, c, alpha=0.3)
+ax1.plot(space[C == -1], R[C == -1], 'k.', alpha=0.3)
+ax1.plot(space, np.ones_like(space) * 0.75, 'k-', alpha=0.5)
+ax1.plot(space, np.ones_like(space) * 0.25, 'k-.', alpha=0.5)
+
+clust = OPTICS(min_samples=9, rejection_ratio=.5)
+
+# Run the fit
+clust.fit(X)
+
+_, labels_025 = clust.extract_dbscan(0.25)
+_, labels_075 = clust.extract_dbscan(0.75)
+
+space = np.r_[0:len(X):1]
+R = clust.reachability_[clust.ordering_]
+C = clust.labels_[clust.ordering_]
+
+plt.figure(figsize=(10, 7))
+G = gridspec.GridSpec(2, 3)
+ax1 = plt.subplot(G[0, :])
+ax1.set_ylabel('Reachability (epsilon distance)')
+ax1.set_title('Reachability Plot')
+ax2 = plt.subplot(G[1, 0])
+ax2.set_title('Automatic Clustering')
+ax3 = plt.subplot(G[1, 1])
+ax3.set_title('Clustering at 0.25 epsilon cut')
+ax4 = plt.subplot(G[1, 2])
+ax4.set_title('Clustering at 0.75 epsilon cut')
+
+# Reachability plot
+color = ['g.', 'r.', 'b.', 'y.', 'c.']
+for k, c in zip(range(1, 7), color):
+    Xk = space[C == k]
+    Rk = R[C == k]
+    ax1.plot(Xk, Rk, c, alpha=0.3)
+ax1.plot(space[C == -1], R[C == -1], 'k.', alpha=0.3)
+ax1.plot(space, np.ones_like(space) * 0.75, 'k-', alpha=0.5)
+ax1.plot(space, np.ones_like(space) * 0.25, 'k-.', alpha=0.5)
+
+# OPTICS
+color = ['g.', 'r.', 'b.', 'y.', 'c.']
+for k, c in zip(range(1, 7), color):
+    Xk = X[clust.labels_ == k]
+    ax2.plot(Xk[:, 0], Xk[:, 1], c, alpha=0.3)
+ax2.plot(X[clust.labels_ == -1, 0], X[clust.labels_ == -1, 1], 'k+', alpha=0.1)
+
+# DBSCAN at 0.25
+color = ['g.', 'm.', 'k.', 'r.', 'b.', 'c.']
```

Comment:
I think that in ax3 and ax4 some of the same specific clusters are found, although the membership is slightly different. The contiguous area under the dashed line of ax1 in the red region does correspond to the red cluster in ax3; there's just more of the 'black tail' that is also under that line, which is why that cluster in ax3 is slightly bigger. Same for the green cluster in ax2 and ax4... both the optics and dbscan extraction are identifying the same cluster, albeit with slightly different membership since the 'auto' extraction is looking at changes in the slope of the reachability graph while the dbscan is just thresholding at a constant value. 

The idea behind the red/blue colors in ax2/ax3 and purple in ax4 is that if you mix red and blue, you get purple. Hence, for optics (auto) and dbscan at 0.25, you get red and blue clusters with slightly different memberships....but for dbscan at 0.75, those clusters 'merge', and that merged cluster is mixed to purple.

I did change the colors of the 'sub clusters' in ax3
 from black and purple to other colors that aren't represented in ax1.

...reading your comment below made me realize that the figure may not have rendered the way I want it to since the update from 1 to 0 based indexing. This is what I'm referring to in the above comment:

![plot_optics](https://user-images.githubusercontent.com/3896161/40397377-a3e8047e-5def-11e8-81f4-50c455175b02.png)